### PR TITLE
Fixes #59 - Added BasicShaderTestSuite

### LINKED ISSUE #59: BasicShaderTestSuite: Create new TestSuite for testing shader functionality
### Description

A test suite should be developed and added to the framework to test all core shader functionality, as this is an area of functionality that is currently missing from the framework.

### Requirements

The project should be up to date with 2024.6 release.

It should include tests that cover the following:
- General shader usage and interaction with GML
- Functionality of all shader-related GML functions:
  - shader_get_name
  - shader_get_uniform
  - shader_get_sampler_index
  - shader_set
  - shader_set_uniform_f
  - shader_set_uniform_f_array
  - shader_set_uniform_f_buffer
  - shader_set_uniform_i
  - shader_set_uniform_i_array
  - shader_set_uniform_matrix
  - shader_set_uniform_matrix_array
  - shader_reset
  - shader_is_compiled
  - shaders_are_supported
  - shader_current
  - shader_enable_corner_id

### Steps To The Task

- [x] Create a new test suite for testing shader functionalities
- [x] Add facts / theories to test the required functions and shader interactions
- [x] Ensure tests are working correctly and in-line with documentation
- [x] Any errors discovered in GameMaker functionality or documentation during development should be reported as issues in the GameMaker-Bugs repository
- [x] Any errors with the framework itself should be reported as issues within this repository

### PR DESCRIPTION
**Pull request by Indibulous**
Created BasicShaderTestSuite, a TestSuite containing 21 facts/testAsyncs that test all the required functionality listed in https://github.com/YoYoGames/GM-TestFramework/issues/59:
 - primitive_drawing
 - shader_is_compiled
 - shader_get_name
 - shaders_are_supported
 - shader_current
 - passthrough_shader
 - shader_get/set_uniform_f
 - shader_set_uniform_f_array
 - shader_set_uniform_f_buffer
 - shader_set_uniform_i
 - shader_set_uniform_i_array
 - shader_get_sampler_index
 - shader_set_uniform_matrix
 - shader_set_uniform_matrix_array
 - shader_enable_corner_id
 - gl_frag_coord/sv_position
 - gl_max_draw_buffers
 - gl_frag_data/sv_target
 - 3d_rendering
 - normals_test
 - gl_front_facing/sv_is_front_face

Added 38 shaders in TestResources/shaders that are used in the tests:
 - 13 in GLSL
 - 13 in GLSLES
 - 12 in HLSL
 
Added a ShaderTests folder in the "datafiles" folder within the project files, containing 64 expected buffer/image files used in the tests' draw buffer comparisons

Added 16 utility functions in Tests/Utils/ShaderTests related to generating, drawing and manipulating vertex data for the tests
 
Modified and added some additional functions to Modules/TestModule/Utils/draw_test_utils to allow testing of multiple render targets at once:
 - start_draw_comparison_ext()
 - end_draw_comparison_ext()

### CODE REVIEW

**Comment by DiasFranciscoA** [comment]
@Indibulous This PR has some conflicts with the main branch please take time to review it before we merge

**Comment by Indibulous** [comment]
Apologies, should be fixed now.

**Review by AdamBorek — APPROVED** [review]
Added some missing commas, but other than that it seems good to me, awesome work 🤓

File: projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml [hunk 1]
```diff
@@ -0,0 +1,1164 @@
...
+		ev_create: function() {
+			// Generate rectangle data to draw
+			rect = new Rect(0, 0, SHADER_TEST_DEFAULT_SIZE, SHADER_TEST_DEFAULT_SIZE);
+		},
+		ev_draw: function() {
+			// Start draw buffer comparison
+			var _test_surface = start_draw_comparison(SHADER_TEST_DEFAULT_SIZE, SHADER_TEST_DEFAULT_SIZE);
+			
+			// Draw rectangle
+			draw_rect(rect)
```
Comment (AdamBorek) [inline projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml]:
```suggestion
			draw_rect(rect);
```

File: projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml [hunk 2]
```diff
@@ -0,0 +1,1164 @@
...
+				var _info = os_get_info()
+				// If info map contains an entry with information about the platform's supported shader language..
+				if (ds_map_exists(_info, "GL_SHADING_LANGUAGE_VERSION"))
+				{
+					// If the shader language entry has a valid value (not undefined or empty string)
+					shader_language = _info[? "GL_SHADING_LANGUAGE_VERSION"];
+					if (!is_undefined(shader_language) && shader_language != "")
+					{
+						// Check that the shader language name doesn't have "GLSL ES" in it (because if it uses any version of GLSL ES then shaders should be supported)
+						assert_string_contains(shader_language, "OpenGL ES")
```
Comment (AdamBorek) [inline projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml]:
```suggestion
						assert_string_contains(shader_language, "OpenGL ES");
```

File: projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml [hunk 3]
```diff
@@ -0,0 +1,1164 @@
...
+			test_shader = pick_shader_for_platform(sh_passthrough_glsles, sh_passthrough_hlsl, sh_passthrough_glsl);
+			// Check that the shader has been compiled
+			verify_shader_compiled(test_shader);
+			
+		},
+		ev_draw: function() {
+			// Start using shader
+			shader_set(test_shader);
+				// Test that shader_current() returns the shader we're currently using
+				assert_true(shader_current() == test_shader, test_current().name +", failed to get current shader")
```
Comment (AdamBorek) [inline projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml]:
```suggestion
				assert_true(shader_current() == test_shader, test_current().name +", failed to get current shader");
```

File: projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml [hunk 4]
```diff
@@ -0,0 +1,1164 @@
...
+			test_shader = pick_shader_for_platform(sh_uniform_i_glsles, sh_uniform_i_hlsl, sh_uniform_i_glsl);
+			// Check that the shader has been compiled
+			verify_shader_compiled(test_shader);
+			
+			// Generate rectangle data to draw
+			rect = new Rect(0, 0, SHADER_TEST_DEFAULT_SIZE, SHADER_TEST_DEFAULT_SIZE);
+			
+			// Get color uniform handle
+			uni_color = shader_get_uniform(test_shader, "color");
+			assert_typeof(uni_color, "number", test_current().name +", failed to get a valid uniform handle")
```
Comment (AdamBorek) [inline projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml]:
```suggestion
			assert_typeof(uni_color, "number", test_current().name +", failed to get a valid uniform handle");
```

**Review by AdamBorek — APPROVED** [review]

**Review by DiasFranciscoA — APPROVED** [review]

**Pull request merged**